Signal/src/environment/PushRegistrationManager.swift

```diff
@@ -0,0 +1,287 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import PromiseKit
+import PushKit
+
+public enum PushRegistrationError: Error {
+    case assertionError(description: String)
+    case pushNotSupported(description: String)
+    case timeout
+}
+
+/**
+ * Singleton used to integrate with push notification services - registration and routing received remote notifications.
+ */
+@objc class PushRegistrationManager: NSObject, PKPushRegistryDelegate {
```

Comment:
I cleaved the push registration logic from the the PushManager. My hope is too eventually merge the remaining PushManager logic into the NotificationManager.